### ISSUE TITLE
Trouble getting version on release create

### ISSUE THREAD

**Issue by Gakk** [PaulHatch/semantic-version#27]
I have a repository with the latest tag `6.1.0` a few weeks ago, and have later implemented `paulhatch/semantic-version` with great joy 😃

1. Pushing commit containing "(MAJOR)" and getting version 7.0.0+15 in build 👍 
2. Pushing commit and getting version 7.0.0+16 in build 👍

My workflow action is triggered on push, pull request and creation of release:

```
on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]
  release:
    types:
      - created
```

My trouble is with the last one:

3. Creating release (and tag) "v7.0.0" on master branch using GitHub UI and getting version 7.0.0+16 again ❗ 
4. Pushing two commits and getting version 7.0.0+18 in build ❓ 
5. Pushing commit and getting version 7.0.1+2 🤔

When creating release (3) i was expecting version 7.0.0 to be returned as the current commit was tagged by "v7.0.0". In the logs for this build i get: `Checking out the ref /usr/bin/git checkout --progress --force refs/tags/v7.0.0` and getting a detached head.

All other builds log `Checking out the ref /usr/bin/git checkout --progress --force -B master refs/remotes/origin/master`.

The trouble is maybe that the master branch is not checked out when triggered from creating release, but rather that the tag is checked out?

I hope I am making sense, and that you understand what might cause this 😄

**Comment by PaulHatch** [comment]
Hi Frode, thank you for reporting these problems.

I am not quite clear here how your versioning is set up, but this action supports either using a prefixed tag like `v1.0.0` or just a plain `1.0.0`. This is configured using the `tag_prefix` parameter and the default is `v`. You reference a tag name `6.1.0`, then `v7.0.0` later, I'm not sure if you're just leaving it off. Also make sure your repo includes history and tags when the actions runs. The default for `actions/checkout` [does not pull the full history](https://github.com/PaulHatch/semantic-version#important-note-regarding-the-checkout-action) and requires `fetch-depth` to be set.

When you tag a commit with a version, that commit will be the last commit with that version. The next commit will have an incremented number, so if you have a tag for v7.0.0, v7.0.1 would be the version of the next commit (unless it included a major/minor pattern in the message of course). So looking at the items you mentioned here:

> 3. Creating release (and tag) "v7.0.0" on master branch using GitHub UI and getting version 7.0.0+16 again ❗

I am not sure where this "again" happens, but this would make sense only if you'd checked out the v7.0.0 tag itself

> 4. Pushing two commits and getting version 7.0.0+18 in build ❓

That doesn't sound right, assuming the was running off the last build. (Check that `v` prefix thing?)

> 5. Pushing commit and getting version 7.0.1+2 🤔

That's the correct version, though the increment is zero based and I'd expect it to be 7.0.1+1 if there are two commits.

It would be extremely helpful if you could clone the repository and create a unit test that replicates the problem you're experiencing and post it here. In the [index.test.js](https://github.com/PaulHatch/semantic-version/blob/master/index.test.js) there is a repo utility that makes creating these tests pretty easy. The repo utility supports these methods:

- `makeCommit(commitMessage[, path])` adds a file, optionally at the specified path, and commits it using a commit message
- `runAction(inputs)` executes the semantic version method and returns the output as a string, specify additional optional inputs to override the defaults
- `exec(command)` runs arbitrary code at repo root
- `clean()` removes the directory containing the temp repo (this is performed automatically when the test is finished)

Here is an example. If you can make this fail for your example when running `npm test` it should be easy to identify or explain what is happening.

```
test('Versioning Test', () => {
    const repo = createTestRepo({ short_tags: 'false' });

    repo.makeCommit('Initial Commit');
    repo.makeCommit('Second Commit');
    repo.makeCommit('Third Commit');
    repo.exec('git tag v6.1.0');
    repo.makeCommit('Fourth Commit (MAJOR)');

    let result = repo.runAction();
    expect(result).toMatch('Version is 7.0.0+0');

    repo.exec("git tag v7.0.0");
    repo.makeCommit('Fifth Commit');

    result = repo.runAction();
    expect(result).toMatch('Version is 7.0.1+0');

    repo.makeCommit('Sixth Commit');

    result = repo.runAction();
    expect(result).toMatch('Version is 7.0.1+1');

    repo.clean();
});
```

**Comment by Gakk** [comment]
I am sorry for not writing clear enough, and for having typos for `6.1.0` when meaning `v6.1.0`.

To better explain what I mean I have created an example repository that you can find at:
https://github.com/Gakk/issue-actionstep-paulhatch-semantic-version

In this repository I have 6 commits and a build action that creates a dummy artifact with the version as name, to quickly see which version was assigned.

Descriptions of commits:
1. Initial adding dummy file. This commit is tagged with `v1.0.0`, expecting and getting version **1.0.0+0** _(build not yet setup, so no artifact og log to see)_
2. Adding workflow, expecting and getting version **1.0.1+0** (except an bug in the workflow naming the artifact only _1.0.1_)
3. Fixing workflow, expecting and getting version **1.0.1+1** 👍 
4. Committing with `(MAJOR)`, expecting and getting version **2.0.0+0** _(even though I wrote in the commit expecting 1.1.0+0 since I meant to write MINOR)_
5. Committing, expecting and getting version **2.0.0+1** 👍 

So far, so good. Everything as expected if you look away from my major/minor typo 😃 

6. I am then creating a release (using GitHub website) with the tag name `v2.0.0` based on commit number 5 that resulted in _2.0.0+1_, because I wanted to release that version.
   - Two workflows start running, probably because I am triggering both `push `and `release:create`, will fix this later.
   - Both workflows ends up with version **2.0.0+1**, even if this commit is tagged with **v2.0.0**. What am I missing? I believed that version tags were the golden master for getting versions.

Pushing the final commit I get versjon 2.0.1+0 as expected again.

What I don't understand is why I get version **2.0.0+1** on on a commit that is tagged with `v2.0.0` 😕

Looks like the logic maybe looks on all previous commits until a version tag is found, and is skipping checking the current commit?

**Comment by PaulHatch** [comment]
The `+1` is the iteration, this not part of the semantic version. This is provided to allow a label such as `-preview1` to be created. Tagging a commit the the expected next version is supposed to preserve that increment.

There was indeed an issue in an older version which would affect running on tagged commits, but this should be fixed as of [v3.2.1](https://github.com/PaulHatch/semantic-version/releases/tag/v3.2.1). [It looks like this repo is using v3.0.0](https://github.com/Gakk/issue-actionstep-paulhatch-semantic-version/blob/master/.github/workflows/main.yml#L27), and as I referenced in the other thread, the latest version is v4.0.1.

**Comment by Gakk** [comment]
Thanks, @PaulHatch. Your answer and a good nights sleep have gotten me on track 😄

Regarding version to use, GitHub specifies in their [documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-versioned-actions) that you can reference a major version:

> : # Reference the major version of a release
> :- uses: actions/setup-node@v1

So since I am using `uses: paulhatch/semantic-version@v3` i believed that would result in the newest version 3.x, meaning 3.2.1. After [enabling step debug logging](https://docs.github.com/en/actions/managing-workflow-runs/enabling-debug-logging) it shows this is not the case 😮

There are multiple other threads online believing specifying just major version implicitly gets newest version within that major, so I believe that GitHub's documentation is a bit weak on that point.

Thanks for your time, and keep up maintaining this great action step 😀👍
